### PR TITLE
Use upstream branch when opening pull requests

### PR DESCRIPTION
**Pull request by mark2185**
- **PR Description**

Should probably solve #2691

- **Please check if the PR fulfills these requirements**

* [x] Cheatsheets are up-to-date (run `go run scripts/cheatsheet/main.go generate`)
* [x] Code has been formatted (see [here](https://github.com/jesseduffield/lazygit/blob/master/CONTRIBUTING.md#code-formatting))
* [x] Tests have been added/updated (see [here](https://github.com/jesseduffield/lazygit/blob/master/pkg/integration/README.md) for the integration test guide)
* [x] Text is internationalised (see [here](https://github.com/jesseduffield/lazygit/blob/master/CONTRIBUTING.md#internationalisation))
* [x] Docs (specifically `docs/Config.md`) have been updated if necessary
* [x] You've read through your own file changes for silly mistakes etc

### CODE REVIEW

**Comment by 12bchl** [comment]
LGTM! My only question would be how it handles a missing upstream name. 

What are your thoughts about this logic flow:

- If upstream name is undefined:
> Input Prompt to set upstream (default value local name)

*Note: git won't let you set an different upstream name that doesn't exist yet on remote. You need to `git push -u origin localname:desiredremotename` to achieve this I believe.

- If input0 or current upstream does not exist on remote:
> Confirmation prompt {Branch xyz does not exist on remote, would you like to push new branch ###} (input0 or current upstream)

- If local branch is out of sync with remote:
> Confirmation prompt {Your local branch xyz is # commits behind/ahead of remote xyz, are you sure you want to open PR?}

Then open PR with upstream name

**Comment by mark2185** [comment]
> What are your thoughts about this logic flow:
> 
>     * If upstream name is undefined:
> 
> 
> > Input Prompt to set upstream (default value local name)
> 
> *Note: git won't let you set an different upstream name that doesn't exist yet on remote. You need to `git push -u origin localname:desiredremotename` to achieve this I believe.


I'd rather throw an error and instruct the user to use `u` to set the branch upstream. If the branch doesn't exist on the remote, pushing the branch through `lazygit` will prompt the user for a name (that's the default setting).

>     * If input0 or current upstream does not exist on remote:
> 
> 
> > Confirmation prompt {Branch xyz does not exist on remote, would you like to push new branch ###} (input0 or current upstream)
> 
>     * If local branch is out of sync with remote:
> 

I'd still use the remote name, it's up to the user to push everything they want, the icons are there by the branch name for that specific reason, to indicate if there's something to push/pull.

**Comment by 12bchl** [comment]
Works for me

**Comment by mark2185** [comment]
@jesseduffield just by perusing the names of the tests, we don't seem to have any.

Should I add a few that test the `error` pop-up since the other ones should open up a browser?

File: pkg/gui/controllers/branches_controller.go [hunk 1]
```diff
@@ -206,7 +206,10 @@ func (self *BranchesController) press(selectedBranch *models.Branch) error {
 }
 
 func (self *BranchesController) handleCreatePullRequest(selectedBranch *models.Branch) error {
-	return self.createPullRequest(selectedBranch.Name, "")
+	if len(selectedBranch.UpstreamBranch) == 0 {
```
Comment (jesseduffield) [inline pkg/gui/controllers/branches_controller.go:209]: Could we use `!selectedBranch.IsTrackingRemote()` here? Likewise below

File: pkg/gui/controllers/branches_controller.go [hunk 2]
```diff
@@ -459,6 +462,9 @@ func (self *BranchesController) createPullRequestMenu(selectedBranch *models.Bra
 			{
 				LabelColumns: fromToLabelColumns(branch.Name, self.c.Tr.DefaultBranch),
 				OnPress: func() error {
+					if len(branch.UpstreamBranch) == 0 {
```
Comment (jesseduffield) [inline pkg/gui/controllers/branches_controller.go:465]: I'd just call `handleCreatePullRequest` from here (We don't want to do `self.createPullRequest(branch.Name, "")` below and looks like we're doing the right thing on line 212

File: pkg/gui/controllers/branches_controller.go [hunk 3]
```diff
@@ -482,6 +495,9 @@ func (self *BranchesController) createPullRequestMenu(selectedBranch *models.Bra
 			&types.MenuItem{
 				LabelColumns: fromToLabelColumns(checkedOutBranch.Name, selectedBranch.Name),
 				OnPress: func() error {
+					if len(checkedOutBranch.UpstreamBranch) == 0 || len(selectedBranch.UpstreamBranch) == 0 {
+						return self.c.ErrorMsg(self.c.Tr.PullRequestNoUpstream)
+					}
 					return self.createPullRequest(checkedOutBranch.Name, selectedBranch.Name)
```
Comment (jesseduffield) [inline pkg/gui/controllers/branches_controller.go:501]: Should we pass the upstream branch names here?

**Comment by mark2185** [comment]
> Sorry for the belated review! I left some comments.
> 
> Yep, some new tests would be good for this

Sorry for the belated followup :)

I can't think of any tests that I could add except trying to open up a PR for a branch without an upstream.

**Comment by jesseduffield** [comment]
Pushed a small change relocating the test. Happy to merge once CI passes :)

**Comment by mark2185** [comment]
The CI passed! :)

(pinging just so it can be closed and forgotten)

**Comment by jesseduffield** [comment]
Thanks for your patience @mark2185 !

**Pull request merged**